Analysis/bin/SimpleMergeMix.cxx

```diff
@@ -0,0 +1,426 @@
+#include "TauMLTools/Core/interface/ProgressReporter.h"
+#include "TauMLTools/Analysis/interface/TauTuple.h"
+#include "TauMLTools/Analysis/interface/GenLepton.h"
+#include "TauMLTools/Analysis/interface/DisTauTagSelection.h"
+
+namespace analysis {
+
+struct IterIdx {
+  size_t file_idx;
+  size_t event_idx;
+  IterIdx () {}
+  IterIdx(const size_t _file_idx, 
+          const size_t _event_idx) 
+          : file_idx(_file_idx),
+            event_idx(_event_idx) {}
+};
+
+class FilesSplit {
+
+  public:
+
+    IterIdx point_entry;
+    IterIdx point_exit;
+    size_t step;
+
+    FilesSplit() {}
+    FilesSplit(const std::vector<size_t>& files_entries,
+               const size_t job_idx,
+               const size_t n_job)
+    {
+      // FilesSplit splits files into n_job sub-intervals
+      // the entry events for the job number job_idx are [point_entry, point_exit]...
+      // if sum(files_entries) % n_job != 0 some events will be lost in datagroup
+      if(job_idx>=n_job) throw exception("Error: job_idx should be < n_job");
+
+      step = std::accumulate(files_entries.begin(), files_entries.end(), 0) / n_job;
+      auto find_idx = [&](const size_t index, const bool isExit) -> IterIdx {
+        size_t sum_accumulate = 0;
+        for(size_t f_i = 0; f_i < files_entries.size(); f_i++) {
+          if(step*(index+isExit) - sum_accumulate - isExit < files_entries[f_i])
+            return IterIdx(f_i, step*(index+isExit) - sum_accumulate  - isExit);
+          sum_accumulate+=files_entries[f_i];
+        }
+        throw exception("Error: sum-based splitting error!");
+      };
+      point_entry = find_idx(job_idx, false);
+      point_exit = find_idx(job_idx, true);
+    }
+};
+
+struct Arguments {
+    run::Argument<std::string> cfg{"cfg", "configuration file with the list of input sources", ""};
+    run::Argument<std::string> input{"input", "Input file with the list of files to read. ", ""};
+    run::Argument<std::string> output{"output", "output", ""};
+    // run::Argument<size_t> file_entries{"file-entries", "maximal number of entries per one file", std::numeric_limits<size_t>::max()};
+    run::Argument<size_t> max_entries{"max-entries", "maximal number of entries processed", std::numeric_limits<size_t>::max()};
+    run::Argument<unsigned> n_threads{"n-threads", "number of threads", 1};
+    run::Argument<unsigned> seed{"seed", "random seed to initialize the generator used for sampling", 1234567};
+    run::Argument<std::string> disabled_branches{"disabled-branches", "list of branches to disabled in the input tuples", ""};
+    run::Argument<std::string> compression_algo{"compression-algo","ZLIB, LZMA, LZ4","LZMA"};
+    run::Argument<unsigned> compression_level{"compression-level", "compression level of output file", 9};
+    run::Argument<unsigned> job_idx{"job-idx", "index of the job (starts from 0)"};
+    run::Argument<unsigned> n_jobs{"n-jobs", "the number by which to divide all files"};
+};
+
+struct SourceDesc {
```

Comment:
there is a lot of code duplication with S&M code. Could the common part be extracted in .h and merging logic is customised with inheritance and virtual functions?

Reply:
No, at the moment I did not add this functionality, for instance DisplacedTau definition does not valid for any analysis but would just require more time to create spectrum hists and also initializing of some extra variables. I propose every analysis carefully checks its own definition.